### ISSUE TITLE
[bitnami/postgresql-ha] ERROR:  cannot execute INSERT in a read-only transaction

### ISSUE THREAD

**Issue by rohitkgupta** [bitnami/charts#23219]
### Name and Version

bitnami/postgresql-ha 6.5.5

### What architecture are you using?

None

### What steps will reproduce the bug?

1. Deploy PostgreSQL with HA architecture (using Replication Manager (repmgr) and Pgpool using bitnami helm charts
Postgresql replicas=2 and Pgpool replicas=2
2. C++ application connected to DB via Pgpool using the libpq library and executing SELECT, INSERT and Update queries.
3. Occasionally, when a master PostgreSQL node switches to slave and slave node to master due to a restart or any other reason, Pgpool may continue to send all queries to the previous master, which is now acting as a slave. This results in the following error:`ERROR: cannot execute INSERT in a read-only transaction`
This error occurs because Pgpool attempts to execute INSERT and UPDATE queries on a PostgreSQL instance that is in a read-only state.

### Are you using any custom parameters or values?
```
----------------------------
pgPool-II configuration file
----------------------------

backend_clustering_mode = 'streaming_replication'

listen_addresses = '*'
port = '5432'
socket_dir = '/opt/build/pgpool/tmp'
reserved_connections = 0



pcp_listen_addresses = '*'
pcp_port = 9898
pcp_socket_dir = '/opt/build/pgpool/tmp'
listen_backlog_multiplier = 2
serialize_accept = off


backend_hostname0 = 'host1'
backend_port0 = 5432
backend_weight0 = 1
backend_data_directory0 = '/data'
backend_flag0 = 'ALLOW_TO_FAILOVER'
backend_application_name0 = 'server0'


enable_pool_hba = 'on'
authentication_timeout = '30'

allow_clear_text_frontend_auth = 'off'


ssl = 'on'
ssl_key = '/opt/build/pgpool/certs/tls.key'
ssl_cert = '/opt/build/pgpool/certs/tls.crt'

ssl_ciphers = 'HIGH:MEDIUM:+3DES:!aNULL'
ssl_prefer_server_ciphers = off
ssl_ecdh_curve = 'prime256v1'
ssl_dh_params_file = ''



num_init_children = '200'
max_pool = '15'


child_life_time = 5min
child_max_connections = 0
connection_life_time = 0
client_idle_limit = 0




log_destination = 'stderr'



log_connections = 'off'
log_disconnections = off
log_hostname = 'on'
log_statement = off
log_per_node_statement = 'off'
log_client_messages = off
log_standby_delay = 'if_over_threshold'


syslog_facility = 'LOCAL0'
syslog_ident = 'pgpool'



pid_file_name = '/opt/build/pgpool/tmp/pgpool.pid'
logdir = '/opt/build/pgpool/logs'



connection_cache = on

reset_query_list = 'ABORT; DISCARD ALL'



replicate_select = off

insert_lock = off
lobj_lock_table = ''


replication_stop_on_mismatch = off

failover_if_affected_tuples_mismatch = off



load_balance_mode = 'off'
ignore_leading_white_space = on
read_only_function_list = ''
write_function_list = ''

primary_routing_query_pattern_list = ''

database_redirect_preference_list = ''

app_name_redirect_preference_list = ''
allow_sql_comments = off

disable_load_balance_on_write = 'transaction'

dml_adaptive_object_relationship_list= ''

statement_level_load_balance = 'off'



sr_check_period = '30'
sr_check_user = 'repmgr'

sr_check_database = 'postgres'
delay_threshold = 10000000


follow_primary_command = ''


health_check_period = '30'
health_check_timeout = '10'
health_check_user = 'repmgr'

health_check_database = ''
health_check_max_retries = '5'
health_check_retry_delay = '5'
connect_timeout = 10000



failover_command = 'echo ">>> Failover - that will initialize new primary node search!"'
failback_command = ''

failover_on_backend_error = 'off'

detach_false_primary = off

search_primary_node_timeout = '0'


recovery_user = 'nobody'
recovery_password = ''

recovery_1st_stage_command = ''
recovery_2nd_stage_command = ''
recovery_timeout = 90
client_idle_limit_in_recovery = 0

auto_failback = off
auto_failback_interval = 1min



use_watchdog = off


trusted_servers = ''
ping_path = '/bin'


hostname0 = ''
wd_port0 = 9000
pgpool_port0 = 9999



wd_priority = 1

wd_authkey = ''

wd_ipc_socket_dir = '/tmp'



delegate_IP = ''
if_cmd_path = '/sbin'
if_up_cmd = '/usr/bin/sudo /sbin/ip addr add $_IP_$/24 dev eth0 label eth0:0'
if_down_cmd = '/usr/bin/sudo /sbin/ip addr del $_IP_$/24 dev eth0'
arping_path = '/usr/sbin'
arping_cmd = '/usr/bin/sudo /usr/sbin/arping -U $_IP_$ -w 1 -I eth0'


clear_memqcache_on_escalation = on
wd_escalation_command = ''
wd_de_escalation_command = ''


failover_when_quorum_exists = on

failover_require_consensus = on

allow_multiple_failover_requests_from_node = off


enable_consensus_with_half_votes = off




wd_lifecheck_method = 'heartbeat'
wd_interval = 10


heartbeat_hostname0 = ''
heartbeat_port0 = 9694
heartbeat_device0 = ''


wd_heartbeat_keepalive = 2
wd_heartbeat_deadtime = 30


wd_life_point = 3
wd_lifecheck_query = 'SELECT 1'
wd_lifecheck_dbname = 'template1'
wd_lifecheck_user = 'nobody'
wd_lifecheck_password = ''

relcache_expire = 0
relcache_size = 256

check_temp_table = catalog

check_unlogged_table = on
enable_shared_relcache = on

memory_cache_enabled = off
memqcache_method = 'shmem'
memqcache_memcached_host = 'localhost'
memqcache_memcached_port = 11211
memqcache_total_size = 64MB
memqcache_max_num_cache = 1000000
memqcache_expire = 0
memqcache_auto_cache_invalidation = on
memqcache_maxcache = 400kB
memqcache_cache_block_size = 1MB
memqcache_oiddir = '/var/log/pgpool/oiddir'
cache_safe_memqcache_table_list = ''
cache_unsafe_memqcache_table_list = ''
backend_hostname0 = 'postgresql-ha-postgresql-0.postgresql-ha-postgresql-headless.ns1.svc.cluster.local'
backend_port0 = 5432
backend_weight0 = 1
backend_data_directory0 = '/opt/build/pgpool/data'
backend_flag0 = 'ALLOW_TO_FAILOVER'
backend_hostname1 = 'postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.ns1.svc.cluster.local'
backend_port1 = 5432
backend_weight1 = 1
backend_data_directory1 = '/opt/build/pgpool/data'
backend_flag1 = 'ALLOW_TO_FAILOVER'
```

### What is the expected behavior?

Pgpool should consistently direct queries to the master node and promptly switch to the new master node whenever a change occurs.

### What do you see instead?

Pgpool doesn't switch to the master node until we restart pgpool deployment.

**Comment by rohitkgupta** [comment]
Postgresql config
-------------------------------
# -----------------------------
# PostgreSQL configuration file
# -----------------------------
#
# This file consists of lines of the form:
#
#   name = value
#
# (The "=" is optional.)  Whitespace may be used.  Comments are introduced with
# "#" anywhere on a line.  The complete list of parameter names and allowed
# values can be found in the PostgreSQL documentation.
#
# The commented-out settings shown in this file represent the default values.
# Re-commenting a setting is NOT sufficient to revert it to the default value;
# you need to reload the server.
#
# This file is read on server startup and when the server receives a SIGHUP
# signal.  If you edit the file on a running system, you have to SIGHUP the
# server for the changes to take effect, run "pg_ctl reload", or execute
# "SELECT pg_reload_conf()".  Some parameters, which are marked below,
# require a server shutdown and restart to take effect.
#
# Any parameter can also be given as a command-line option to the server, e.g.,
# "postgres -c log_connections=on".  Some parameters can be changed at run time
# with the "SET" SQL command.
#
# Memory units:  B  = bytes            Time units:  us  = microseconds
#                kB = kilobytes                     ms  = milliseconds
#                MB = megabytes                     s   = seconds
#                GB = gigabytes                     min = minutes
#                TB = terabytes                     h   = hours
#                                                   d   = days


#------------------------------------------------------------------------------
# FILE LOCATIONS
#------------------------------------------------------------------------------

# The default values of these variables are driven from the -D command-line
# option or PGDATA environment variable, represented here as ConfigDir.

#data_directory = 'ConfigDir'		# use data in another directory
					# (change requires restart)
#hba_file = 'ConfigDir/pg_hba.conf'	# host-based authentication file
					# (change requires restart)
#ident_file = 'ConfigDir/pg_ident.conf'	# ident configuration file
					# (change requires restart)

# If external_pid_file is not explicitly set, no extra PID file is written.
#external_pid_file = ''			# write an extra PID file
					# (change requires restart)


#------------------------------------------------------------------------------
# CONNECTIONS AND AUTHENTICATION
#------------------------------------------------------------------------------

# - Connection Settings -

listen_addresses = '*'
					# comma-separated list of addresses;
					# defaults to 'localhost'; use '*' for all
					# (change requires restart)
port = '5432'
max_connections = '512'
#superuser_reserved_connections = 3	# (change requires restart)
#unix_socket_directories = '/tmp'	# comma-separated list of directories
					# (change requires restart)
#unix_socket_group = ''			# (change requires restart)
#unix_socket_permissions = 0777		# begin with 0 to use octal notation
					# (change requires restart)
#bonjour = off				# advertise server via Bonjour
					# (change requires restart)
#bonjour_name = ''			# defaults to the computer name
					# (change requires restart)

# - TCP settings -
# see "man tcp" for details

#tcp_keepalives_idle = 0		# TCP_KEEPIDLE, in seconds;
					# 0 selects the system default
#tcp_keepalives_interval = 0		# TCP_KEEPINTVL, in seconds;
					# 0 selects the system default
#tcp_keepalives_count = 0		# TCP_KEEPCNT;
					# 0 selects the system default
#tcp_user_timeout = 0			# TCP_USER_TIMEOUT, in milliseconds;
					# 0 selects the system default

# - Authentication -

#authentication_timeout = 1min		# 1s-600s
#password_encryption = md5		# md5 or scram-sha-256
#db_user_namespace = off

# GSSAPI using Kerberos
#krb_server_keyfile = 'FILE:${sysconfdir}/krb5.keytab'
#krb_caseins_users = off

# - SSL -

ssl = 'on'
#ssl_ca_file = ''
ssl_cert_file = '/certs/tls.crt'
#ssl_crl_file = ''
ssl_key_file = '/certs/tls.key'
#ssl_ciphers = 'HIGH:MEDIUM:+3DES:!aNULL' # allowed SSL ciphers
#ssl_prefer_server_ciphers = on
#ssl_ecdh_curve = 'prime256v1'
#ssl_min_protocol_version = 'TLSv1.2'
#ssl_max_protocol_version = ''
#ssl_dh_params_file = ''
#ssl_passphrase_command = ''
#ssl_passphrase_command_supports_reload = off


#------------------------------------------------------------------------------
# RESOURCE USAGE (except WAL)
#------------------------------------------------------------------------------

# - Memory -

#shared_buffers = 32MB			# min 128kB
					# (change requires restart)
#huge_pages = try			# on, off, or try
					# (change requires restart)
#temp_buffers = 8MB			# min 800kB
#max_prepared_transactions = 0		# zero disables the feature
					# (change requires restart)
# Caution: it is not advisable to set max_prepared_transactions nonzero unless
# you actively intend to use prepared transactions.
#work_mem = 4MB				# min 64kB
#hash_mem_multiplier = 1.0		# 1-1000.0 multiplier on hash table work_mem
#maintenance_work_mem = 64MB		# min 1MB
#autovacuum_work_mem = -1		# min 1MB, or -1 to use maintenance_work_mem
#logical_decoding_work_mem = 64MB	# min 64kB
#max_stack_depth = 2MB			# min 100kB
#shared_memory_type = mmap		# the default is the first option
					# supported by the operating system:
					#   mmap
					#   sysv
					#   windows
					# (change requires restart)
#dynamic_shared_memory_type = posix	# the default is the first option
					# supported by the operating system:
					#   posix
					#   sysv
					#   windows
					#   mmap
					# (change requires restart)

# - Disk -

#temp_file_limit = -1			# limits per-process temp file space
					# in kilobytes, or -1 for no limit

# - Kernel Resources -

#max_files_per_process = 1000		# min 64
					# (change requires restart)

# - Cost-Based Vacuum Delay -

#vacuum_cost_delay = 0			# 0-100 milliseconds (0 disables)
#vacuum_cost_page_hit = 1		# 0-10000 credits
#vacuum_cost_page_miss = 10		# 0-10000 credits
#vacuum_cost_page_dirty = 20		# 0-10000 credits
#vacuum_cost_limit = 200		# 1-10000 credits

# - Background Writer -

#bgwriter_delay = 200ms			# 10-10000ms between rounds
#bgwriter_lru_maxpages = 100		# max buffers written/round, 0 disables
#bgwriter_lru_multiplier = 2.0		# 0-10.0 multiplier on buffers scanned/round
#bgwriter_flush_after = 0		# measured in pages, 0 disables

# - Asynchronous Behavior -

#effective_io_concurrency = 1		# 1-1000; 0 disables prefetching
#maintenance_io_concurrency = 10	# 1-1000; 0 disables prefetching
#max_worker_processes = 8		# (change requires restart)
#max_parallel_maintenance_workers = 2	# taken from max_parallel_workers
#max_parallel_workers_per_gather = 2	# taken from max_parallel_workers
#parallel_leader_participation = on
#max_parallel_workers = 8		# maximum number of max_worker_processes that
					# can be used in parallel operations
#old_snapshot_threshold = -1		# 1min-60d; -1 disables; 0 is immediate
					# (change requires restart)
#backend_flush_after = 0		# measured in pages, 0 disables


#------------------------------------------------------------------------------
# WRITE-AHEAD LOG
#------------------------------------------------------------------------------

# - Settings -

wal_level = 'hot_standby'
					# (change requires restart)
fsync = 'on'
					# (turning this off can cause
					# unrecoverable data corruption)
#synchronous_commit = on		# synchronization level;
					# off, local, remote_write, remote_apply, or on
#wal_sync_method = fsync		# the default is the first option
					# supported by the operating system:
					#   open_datasync
					#   fdatasync (default on Linux and FreeBSD)
					#   fsync
					#   fsync_writethrough
					#   open_sync
#full_page_writes = on			# recover from partial page writes
#wal_compression = off			# enable compression of full-page writes
#wal_log_hints = off			# also do full page writes of non-critical updates
					# (change requires restart)
#wal_init_zero = on			# zero-fill new WAL files
#wal_recycle = on			# recycle WAL files
#wal_buffers = -1			# min 32kB, -1 sets based on shared_buffers
					# (change requires restart)
#wal_writer_delay = 200ms		# 1-10000 milliseconds
#wal_writer_flush_after = 1MB		# measured in pages, 0 disables
#wal_skip_threshold = 2MB

#commit_delay = 0			# range 0-100000, in microseconds
#commit_siblings = 5			# range 1-1000

# - Checkpoints -

#checkpoint_timeout = 5min		# range 30s-1d
max_wal_size = '400MB'
#min_wal_size = 80MB
#checkpoint_completion_target = 0.5	# checkpoint target duration, 0.0 - 1.0
#checkpoint_flush_after = 0		# measured in pages, 0 disables
#checkpoint_warning = 30s		# 0 disables

# - Archiving -

archive_mode = 'on'
				# (change requires restart)
archive_command = '/bin/true'
				# placeholders: %p = path of file to archive
				#               %f = file name only
				# e.g. 'test ! -f /mnt/server/archivedir/%f && cp %p /mnt/server/archivedir/%f'
#archive_timeout = 0		# force a logfile segment switch after this
				# number of seconds; 0 disables

# - Archive Recovery -

# These are only used in recovery mode.

#restore_command = ''		# command to use to restore an archived logfile segment
				# placeholders: %p = path of file to restore
				#               %f = file name only
				# e.g. 'cp /mnt/server/archivedir/%f %p'
				# (change requires restart)
#archive_cleanup_command = ''	# command to execute at every restartpoint
#recovery_end_command = ''	# command to execute at completion of recovery

# - Recovery Target -

# Set these only when performing a targeted recovery.

#recovery_target = ''		# 'immediate' to end recovery as soon as a
                                # consistent state is reached
				# (change requires restart)
#recovery_target_name = ''	# the named restore point to which recovery will proceed
				# (change requires restart)
#recovery_target_time = ''	# the time stamp up to which recovery will proceed
				# (change requires restart)
#recovery_target_xid = ''	# the transaction ID up to which recovery will proceed
				# (change requires restart)
#recovery_target_lsn = ''	# the WAL LSN up to which recovery will proceed
				# (change requires restart)
#recovery_target_inclusive = on # Specifies whether to stop:
				# just after the specified recovery target (on)
				# just before the recovery target (off)
				# (change requires restart)
#recovery_target_timeline = 'latest'	# 'current', 'latest', or timeline ID
				# (change requires restart)
#recovery_target_action = 'pause'	# 'pause', 'promote', 'shutdown'
				# (change requires restart)


#------------------------------------------------------------------------------
# REPLICATION
#------------------------------------------------------------------------------

# - Sending Servers -

# Set these on the master and on any standby that will send replication data.

max_wal_senders = '16'
				# (change requires restart)
wal_keep_size = '128MB'
#max_slot_wal_keep_size = -1	# in megabytes; -1 disables
#wal_sender_timeout = 60s	# in milliseconds; 0 disables

max_replication_slots = '10'
				# (change requires restart)
#track_commit_timestamp = off	# collect timestamp of transaction commit
				# (change requires restart)

# - Master Server -

# These settings are ignored on a standby server.

#synchronous_standby_names = ''	# standby servers that provide sync rep
				# method to choose sync standbys, number of sync standbys,
				# and comma-separated list of application_name
				# from standby(s); '*' = all
#vacuum_defer_cleanup_age = 0	# number of xacts by which cleanup is delayed

# - Standby Servers -

# These settings are ignored on a master server.

primary_conninfo = 'host=postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.ns1.svc.cluster.local port=5432 user=repmgr password=aEriXghwbxOjLP3uD8VhIKNHvjbaMO7k application_name=walreceiver'
#primary_slot_name = ''			# replication slot on sending server
promote_trigger_file = '/tmp/postgresql.trigger.5432'
hot_standby = 'on'
					# (change requires restart)
#max_standby_archive_delay = 30s	# max delay before canceling queries
					# when reading WAL from archive;
					# -1 allows indefinite delay
#max_standby_streaming_delay = 30s	# max delay before canceling queries
					# when reading streaming WAL;
					# -1 allows indefinite delay
#wal_receiver_create_temp_slot = off	# create temp slot if primary_slot_name
					# is not set
#wal_receiver_status_interval = 10s	# send replies at least this often
					# 0 disables
#hot_standby_feedback = off		# send info from standby to prevent
					# query conflicts
#wal_receiver_timeout = 60s		# time that receiver waits for
					# communication from master
					# in milliseconds; 0 disables
#wal_retrieve_retry_interval = 5s	# time to wait before retrying to
					# retrieve WAL after a failed attempt
#recovery_min_apply_delay = 0		# minimum delay for applying changes during recovery

# - Subscribers -

# These settings are ignored on a publisher.

#max_logical_replication_workers = 4	# taken from max_worker_processes
					# (change requires restart)
#max_sync_workers_per_subscription = 2	# taken from max_logical_replication_workers


#------------------------------------------------------------------------------
# QUERY TUNING
#------------------------------------------------------------------------------

# - Planner Method Configuration -

#enable_bitmapscan = on
#enable_hashagg = on
#enable_hashjoin = on
#enable_indexscan = on
#enable_indexonlyscan = on
#enable_material = on
#enable_mergejoin = on
#enable_nestloop = on
#enable_parallel_append = on
#enable_seqscan = on
#enable_sort = on
#enable_incremental_sort = on
#enable_tidscan = on
#enable_partitionwise_join = off
#enable_partitionwise_aggregate = off
#enable_parallel_hash = on
#enable_partition_pruning = on

# - Planner Cost Constants -

#seq_page_cost = 1.0			# measured on an arbitrary scale
#random_page_cost = 4.0			# same scale as above
#cpu_tuple_cost = 0.01			# same scale as above
#cpu_index_tuple_cost = 0.005		# same scale as above
#cpu_operator_cost = 0.0025		# same scale as above
#parallel_tuple_cost = 0.1		# same scale as above
#parallel_setup_cost = 1000.0	# same scale as above

#jit_above_cost = 100000		# perform JIT compilation if available
					# and query more expensive than this;
					# -1 disables
#jit_inline_above_cost = 500000		# inline small functions if query is
					# more expensive than this; -1 disables
#jit_optimize_above_cost = 500000	# use expensive JIT optimizations if
					# query is more expensive than this;
					# -1 disables

#min_parallel_table_scan_size = 8MB
#min_parallel_index_scan_size = 512kB
#effective_cache_size = 4GB

# - Genetic Query Optimizer -

#geqo = on
#geqo_threshold = 12
#geqo_effort = 5			# range 1-10
#geqo_pool_size = 0			# selects default based on effort
#geqo_generations = 0			# selects default based on effort
#geqo_selection_bias = 2.0		# range 1.5-2.0
#geqo_seed = 0.0			# range 0.0-1.0

# - Other Planner Options -

#default_statistics_target = 100	# range 1-10000
#constraint_exclusion = partition	# on, off, or partition
#cursor_tuple_fraction = 0.1		# range 0.0-1.0
#from_collapse_limit = 8
#join_collapse_limit = 8		# 1 disables collapsing of explicit
					# JOIN clauses
#force_parallel_mode = off
#jit = on				# allow JIT compilation
#plan_cache_mode = auto			# auto, force_generic_plan or
					# force_custom_plan


#------------------------------------------------------------------------------
# REPORTING AND LOGGING
#------------------------------------------------------------------------------

# - Where to Log -

#log_destination = 'stderr'		# Valid values are combinations of
					# stderr, csvlog, syslog, and eventlog,
					# depending on platform.  csvlog
					# requires logging_collector to be on.

# This is used when logging to stderr:
logging_collector = 'on'
					# into log files. Required to be on for
					# csvlogs.
					# (change requires restart)

# These are only used if logging_collector is on:
log_directory = '/opt/build/postgresql/logs'
					# can be absolute or relative to PGDATA
log_filename = 'postgresql.log'
					# can include strftime() escapes
#log_file_mode = 0600			# creation mode for log files,
					# begin with 0 to use octal notation
#log_truncate_on_rotation = off		# If on, an existing log file with the
					# same name as the new log file will be
					# truncated rather than appended to.
					# But such truncation only occurs on
					# time-driven rotation, not on restarts
					# or size-driven rotation.  Default is
					# off, meaning append to existing files
					# in all cases.
#log_rotation_age = 1d			# Automatic rotation of logfiles will
					# happen after that time.  0 disables.
#log_rotation_size = 10MB		# Automatic rotation of logfiles will
					# happen after that much log output.
					# 0 disables.

# These are relevant when logging to syslog:
#syslog_facility = 'LOCAL0'
#syslog_ident = 'postgres'
#syslog_sequence_numbers = on
#syslog_split_messages = on

# This is only relevant when logging to eventlog (win32):
# (change requires restart)
#event_source = 'PostgreSQL'

# - When to Log -

#log_min_messages = warning		# values in order of decreasing detail:
					#   debug5
					#   debug4
					#   debug3
					#   debug2
					#   debug1
					#   info
					#   notice
					#   warning
					#   error
					#   log
					#   fatal
					#   panic

#log_min_error_statement = error	# values in order of decreasing detail:
					#   debug5
					#   debug4
					#   debug3
					#   debug2
					#   debug1
					#   info
					#   notice
					#   warning
					#   error
					#   log
					#   fatal
					#   panic (effectively off)

#log_min_duration_statement = -1	# -1 is disabled, 0 logs all statements
					# and their durations, > 0 logs only
					# statements running at least this number
					# of milliseconds

#log_min_duration_sample = -1		# -1 is disabled, 0 logs a sample of statements
					# and their durations, > 0 logs only a sample of
					# statements running at least this number
					# of milliseconds;
					# sample fraction is determined by log_statement_sample_rate

#log_statement_sample_rate = 1.0	# fraction of logged statements exceeding
					# log_min_duration_sample to be logged;
					# 1.0 logs all such statements, 0.0 never logs


#log_transaction_sample_rate = 0.0	# fraction of transactions whose statements
					# are logged regardless of their duration; 1.0 logs all
					# statements from all transactions, 0.0 never logs

# - What to Log -

#debug_print_parse = off
#debug_print_rewritten = off
#debug_print_plan = off
#debug_pretty_print = on
#log_checkpoints = off
log_connections = 'false'
log_disconnections = 'false'
#log_duration = off
#log_error_verbosity = default		# terse, default, or verbose messages
log_hostname = 'true'
#log_line_prefix = '%m [%p] '		# special values:
					#   %a = application name
					#   %u = user name
					#   %d = database name
					#   %r = remote host and port
					#   %h = remote host
					#   %b = backend type
					#   %p = process ID
					#   %t = timestamp without milliseconds
					#   %m = timestamp with milliseconds
					#   %n = timestamp with milliseconds (as a Unix epoch)
					#   %i = command tag
					#   %e = SQL state
					#   %c = session ID
					#   %l = session line number
					#   %s = session start timestamp
					#   %v = virtual transaction ID
					#   %x = transaction ID (0 if none)
					#   %q = stop here in non-session
					#        processes
					#   %% = '%'
					# e.g. '<%u%%%d> '
#log_lock_waits = off			# log lock waits >= deadlock_timeout
#log_parameter_max_length = -1		# when logging statements, limit logged
					# bind-parameter values to N bytes;
					# -1 means print in full, 0 disables
#log_parameter_max_length_on_error = 0	# when logging an error, limit logged
					# bind-parameter values to N bytes;
					# -1 means print in full, 0 disables
#log_statement = 'none'			# none, ddl, mod, all
#log_replication_commands = off
#log_temp_files = -1			# log temporary files equal or larger
					# than the specified size in kilobytes;
					# -1 disables, 0 logs all temp files
#log_timezone = 'GMT'

#------------------------------------------------------------------------------
# PROCESS TITLE
#------------------------------------------------------------------------------

#cluster_name = ''			# added to process titles if nonempty
					# (change requires restart)
#update_process_title = on


#------------------------------------------------------------------------------
# STATISTICS
#------------------------------------------------------------------------------

# - Query and Index Statistics Collector -

#track_activities = on
#track_counts = on
#track_io_timing = off
#track_functions = none			# none, pl, all
#track_activity_query_size = 1024	# (change requires restart)
#stats_temp_directory = 'pg_stat_tmp'


# - Monitoring -

#log_parser_stats = off
#log_planner_stats = off
#log_executor_stats = off
#log_statement_stats = off


#------------------------------------------------------------------------------
# AUTOVACUUM
#------------------------------------------------------------------------------

#autovacuum = on			# Enable autovacuum subprocess?  'on'
					# requires track_counts to also be on.
#log_autovacuum_min_duration = -1	# -1 disables, 0 logs all actions and
					# their durations, > 0 logs only
					# actions running at least this number
					# of milliseconds.
#autovacuum_max_workers = 3		# max number of autovacuum subprocesses
					# (change requires restart)
#autovacuum_naptime = 1min		# time between autovacuum runs
#autovacuum_vacuum_threshold = 50	# min number of row updates before
					# vacuum
#autovacuum_vacuum_insert_threshold = 1000	# min number of row inserts
					# before vacuum; -1 disables insert
					# vacuums
#autovacuum_analyze_threshold = 50	# min number of row updates before
					# analyze
#autovacuum_vacuum_scale_factor = 0.2	# fraction of table size before vacuum
#autovacuum_vacuum_insert_scale_factor = 0.2	# fraction of inserts over table
					# size before insert vacuum
#autovacuum_analyze_scale_factor = 0.1	# fraction of table size before analyze
#autovacuum_freeze_max_age = 200000000	# maximum XID age before forced vacuum
					# (change requires restart)
#autovacuum_multixact_freeze_max_age = 400000000	# maximum multixact age
					# before forced vacuum
					# (change requires restart)
#autovacuum_vacuum_cost_delay = 2ms	# default vacuum cost delay for
					# autovacuum, in milliseconds;
					# -1 means use vacuum_cost_delay
#autovacuum_vacuum_cost_limit = -1	# default vacuum cost limit for
					# autovacuum, -1 means use
					# vacuum_cost_limit


#------------------------------------------------------------------------------
# CLIENT CONNECTION DEFAULTS
#------------------------------------------------------------------------------

# - Statement Behavior -

client_min_messages = 'error'
					#   debug5
					#   debug4
					#   debug3
					#   debug2
					#   debug1
					#   log
					#   notice
					#   warning
					#   error
#search_path = '"$user", public'	# schema names
#row_security = on
#default_tablespace = ''		# a tablespace name, '' uses the default
#temp_tablespaces = ''			# a list of tablespace names, '' uses
					# only default tablespace
#default_table_access_method = 'heap'
#check_function_bodies = on
#default_transaction_isolation = 'read committed'
#default_transaction_read_only = off
#default_transaction_deferrable = off
#session_replication_role = 'origin'
#statement_timeout = 0			# in milliseconds, 0 is disabled
#lock_timeout = 0			# in milliseconds, 0 is disabled
#idle_in_transaction_session_timeout = 0	# in milliseconds, 0 is disabled
#vacuum_freeze_min_age = 50000000
#vacuum_freeze_table_age = 150000000
#vacuum_multixact_freeze_min_age = 5000000
#vacuum_multixact_freeze_table_age = 150000000
#vacuum_cleanup_index_scale_factor = 0.1	# fraction of total number of tuples
						# before index cleanup, 0 always performs
						# index cleanup
#bytea_output = 'hex'			# hex, escape
#xmlbinary = 'base64'
#xmloption = 'content'
#gin_fuzzy_search_limit = 0
#gin_pending_list_limit = 4MB

# - Locale and Formatting -

#datestyle = 'iso, mdy'
#intervalstyle = 'postgres'
#timezone = 'GMT'
#timezone_abbreviations = 'Default'     # Select the set of available time zone
					# abbreviations.  Currently, there are
					#   Default
					#   Australia (historical usage)
					#   India
					# You can create your own file in
					# share/timezonesets/.
#extra_float_digits = 1			# min -15, max 3; any value >0 actually
					# selects precise output mode
#client_encoding = sql_ascii		# actually, defaults to database
					# encoding

# These settings are initialized by initdb, but they can be changed.
#lc_messages = 'C'			# locale for system error message
					# strings
#lc_monetary = 'C'			# locale for monetary formatting
#lc_numeric = 'C'			# locale for number formatting
#lc_time = 'C'				# locale for time formatting

# default configuration for text search
#default_text_search_config = 'pg_catalog.simple'

# - Shared Library Preloading -

shared_preload_libraries = 'repmgr, pgaudit, repmgr'
#local_preload_libraries = ''
#session_preload_libraries = ''
#jit_provider = 'llvmjit'		# JIT library to use

# - Other Defaults -

#dynamic_library_path = '$libdir'
#extension_destdir = ''			# prepend path when loading extensions
					# and shared objects (added by Debian)


#------------------------------------------------------------------------------
# LOCK MANAGEMENT
#------------------------------------------------------------------------------

#deadlock_timeout = 1s
#max_locks_per_transaction = 64		# min 10
					# (change requires restart)
#max_pred_locks_per_transaction = 64	# min 10
					# (change requires restart)
#max_pred_locks_per_relation = -2	# negative values mean
					# (max_pred_locks_per_transaction
					#  / -max_pred_locks_per_relation) - 1
#max_pred_locks_per_page = 2            # min 0


#------------------------------------------------------------------------------
# VERSION AND PLATFORM COMPATIBILITY
#------------------------------------------------------------------------------

# - Previous PostgreSQL Versions -

#array_nulls = on
#backslash_quote = safe_encoding	# on, off, or safe_encoding
#escape_string_warning = on
#lo_compat_privileges = off
#operator_precedence_warning = off
#quote_all_identifiers = off
#standard_conforming_strings = on
#synchronize_seqscans = on

# - Other Platforms and Clients -

#transform_null_equals = off


#------------------------------------------------------------------------------
# ERROR HANDLING
#------------------------------------------------------------------------------

#exit_on_error = off			# terminate session on any error?
#restart_after_crash = on		# reinitialize after backend crash?
#data_sync_retry = off			# retry or panic on failure to fsync
					# data?
					# (change requires restart)


#------------------------------------------------------------------------------
# CONFIG FILE INCLUDES
#------------------------------------------------------------------------------

# These options allow settings to be loaded from files other than the
# default postgresql.conf.  Note that these are directives, not variable
# assignments, so they can usefully be given more than once.

include_dir = 'conf.d'
#include_dir = '...'			# include files ending in '.conf' from
					# a directory, e.g., 'conf.d'
#include_if_exists = '...'		# include file only if it exists
#include = '...'			# include file


#------------------------------------------------------------------------------
# CUSTOMIZED OPTIONS
#------------------------------------------------------------------------------

# Add settings for extensions here
pgaudit.log_catalog = 'off'

**Comment by rohitkgupta** [comment]
ping

**Comment by migruiz4** [comment]
Hi @rohitkgupta,

I'm sorry for the late response.

I may need more information to help you with this issue. Does your issue persist over time when the master/slave switch happens? Or is this issue transient?

Please notice that short downtimes may happen when master is switched because in case of failure, the chart behavior would be the following:
- Repmgr will notice the Master node is not available and select a new one.
- Pgpool-II health_check will detect the master node is not available and stop accepting new requests.
- Because of new request requests being refused, first the readiness probe will fail and pgpool-II will no longer receive traffic.
- Eventually, pgpool-II will be restarted by the liveness probe failure, connecting to the new master node.

If the error `ERROR: cannot execute INSERT in a read-only transaction` is indeed transient, I would recommend fine-tunning both the pgpool-II healthcheck settings and the chart liveness/readiness probe settings for quicker recovery from failure.

**Comment by rohitkgupta** [comment]
@migruiz4 
Thanks for your response.
It is not a transient issue. Pgpool didn't restart and there were no probe failure for many hours, we had to restart Pgpool manually in this case.

**Comment by BK-STAR77** [comment]
same issues for me with pgpool:4.5.1-debian-12-r0

**Comment by migruiz4** [comment]
Thank you very much for your feedback @BK-STAR77!

The container logic has not received any change from 4.5.0 to 4.5.1 that could cause this behavior. Checking [pgpool release notes](https://www.pgpool.net/docs/latest/en/html/release-4-5-1.html), it appears that it could be related to the change of the default value for [failover_on_backend_shutdown](https://www.pgpool.net/docs/latest/en/html/runtime-config-failover.html#GUC-FAILOVER-ON-BACKEND-SHUTDOWN).

Could you please confirm if setting `failover_on_backend_shutdown=on` fixes the issue? If that is the case, we should set it as the default value because that may be desired for Kubernetes.

**Comment by apoca** [comment]
I have same issue here and reverting manually to 4.5.0-debian-12-r10 in deployment is working now... but, this could be very dangerous for production...

**Comment by apoca** [comment]
I don't see this variable in values.yaml :/

**Comment by Stevenpc3** [comment]
We are experiencing very bad failover issues also.

We updated to the newer charts from waaaaaay back and the latest charts appear to not failover correctly at all.  We are using kubernetes.

We deploy then deploy our other pods to connect and all is fine.
Next we delete the primary node pod using kubectl delete pod <primary>
everything "appears" fine, but then we delete our other pods connected and they start a db-migration using liquibase and we get this error.

![image](https://github.com/bitnami/charts/assets/6076207/d0a5bcb0-95f4-4026-960b-e2c339d6eaab)

with a corresponding error in pgpool of the following

![image](https://github.com/bitnami/charts/assets/6076207/c731e1ac-fc6c-4684-aedd-b84ba3ad4e20)

There have been other errors two that state it is in read-only mode

`Caused by: org.postgresql.util.PSQLException: ERROR: cannot execute UPDATE in a read-only transaction`

We have attempted to downgrad the image of pgpool to 4.5.0 while still using the current charts but pgpool will fail to start with the following error.

`cp: cannot stat '/opt/bitnami/pgpool/etc/pgpool.conf.sample': No such file or directory`

So currently, it appears the pg 16 chart version 14+ do not properly failover which makes HA useless and dangerous.

We were hoping to update since previous issues with pgpool and the chart caused pgpool to restart beacause of liveness probes and those updates to the probe script were added and we hoped all was fixed.  There was also a previous issue where the DB would go into split brain pretty often and that was super dangerous because it wouldn't fail and you would have one outdated db and one current and it would read from both... so we had to shut off loadbalancing.

**Comment by apoca** [comment]
@Stevenpc3 Exactly the same thing is happening to me, but as I only started using postgresql-ha about a month ago I thought it was a problem with some bad configuration on my part. But I had this problem for weeks, and only by deleting the extra primary one did it start to work, because it recreated everything, but automatically it doesn't work. In my case I ended up having different random data and I noticed that I had two primary.

**Comment by Stevenpc3** [comment]
@apoca that issue you are describing is split-brain https://github.com/bitnami/charts/issues/20998 more on the internet about it too.  The issue I mentioned at the bottom of my comment is more like replication stops working with no errors or issues detected and is super dangerous and makes failover pointless.

Both of those above issues are not related to the current issue in this thread though.  This thread appears to be an issue with PGPool and/or the bitnami chart configs that use it.

**Comment by Stevenpc3** [comment]
You can add the flag 

postgresql-ha:
  pgpool:
    configuration: |
      connection_cache = false
      failover_on_backend_error = on
      failover_on_backend_shutdown = on

**Comment by RaddishIoW** [comment]
> You can add the flag
> 
> ```
> postgresql-ha:
>   pgpool:
>     configuration: |
>       connection_cache = false
>       failover_on_backend_error = on
>       failover_on_backend_shutdown = on
> ```

Is the "postgresql-ha:" name correct? My config file has the name as "postgresql:" - also, the `pgpool:` section is a top-level item in the [values.yaml](https://github.com/bitnami/charts/blob/main/bitnami/postgresql-ha/values.yaml)... Sorry if I'm missing something here - I've just started using this.

Is it not meant to be:
```
postgresql:
...
pgpool:
  configuration: |-
    connection_cache = false
    failover_on_backend_error = on
    failover_on_backend_shutdown = on
```
?

**Comment by Stevenpc3** [comment]
My apologies, your way is correct for using the chart directly. I am using a wrapper chart. We have a chart that uses this chart as a dependency so we can template a few things.

**Comment by apoca** [comment]
Since you talk about this, I have a question here... I have software as a service and I create a schema database and a user/password as well as the necessary permissions in real time... Sometimes I have a problem with Kind because saying that the database does not exist (but it exists) my question is whether this “false” cache solves this type of problem in pgpool?

**Comment by phntom** [comment]
same issue here vanilla install postgresql-ha-14.0.2    16.2.0

**Comment by migruiz4** [comment]
Hi there! 
We have released a new version of the `bitnami/pgpool` container setting the default value for `failover_on_backend_shutdown=on` (https://github.com/bitnami/containers/pull/65526).

This patched image is included in the postgresql-ha Helm chart starting version `14.0.6`. Could you please confirm if the issue persists after upgrading the chart to used the patched pgpool image?

**Comment by apoca** [comment]
@migruiz4 I am using the last version `14.0.10` and I had this error today, and after rollout again pgpool it's working again... This is very serious issue and I won't put this for production mode.

**Comment by migruiz4** [comment]
@apoca Could you please provide a detailed description of your issue? We need additional information in order to troubleshoot and fix this issue

**Comment by Stevenpc3** [comment]
We also tried setting these settings manually and could not get proper fail over to work. I will be doing more testing tomorrow of the charts and try to write a more detailed example of the setup, test, and results. 

What is the fail over testing Bitnami has done to test this? 

It would be nice for us to know and replicate the procedure used to validate the chart and operation to compare.

The basic test we tried and others seem to try is basically... Deploy the chart with 3 instances. Have a service connect through pgpool that does read and write operations. Then delete the primary node pod. Expect a fail over and then the service should still be able to use the database with the new primary... Instead the service can't because pgpool didn't fail over it if it did, it thinks the new primary is read only.

**Comment by Stevenpc3** [comment]
> Hi @rohitkgupta,
> 
> I'm sorry for the late response.
> 
> I may need more information to help you with this issue. Does your issue persist over time when the master/slave switch happens? Or is this issue transient?
> 
> Please notice that short downtimes may happen when master is switched because in case of failure, the chart behavior would be the following:
> 
> * Repmgr will notice the Master node is not available and select a new one.
> * Pgpool-II health_check will detect the master node is not available and stop accepting new requests.
> * Because of new request requests being refused, first the readiness probe will fail and pgpool-II will no longer receive traffic.
> * Eventually, pgpool-II will be restarted by the liveness probe failure, connecting to the new master node.
> 
> If the error `ERROR: cannot execute INSERT in a read-only transaction` is indeed transient, I would recommend fine-tunning both the pgpool-II healthcheck settings and the chart liveness/readiness probe settings for quicker recovery from failure.

From this description, would you state that you EXPECT pgpool to fail liveness probe and subsequently restart as part of a normal and proper fail over?

We have a case where we do not want pgpool to restart. I would hope it was possible for a fail over without restart of pgpool.

**Comment by migruiz4** [comment]
Hi @Stevenpc3, the tests we execute for postgresql-ha can be found here: https://github.com/bitnami/charts/tree/main/.vib/postgresql-ha.

> From this description, would you state that you EXPECT pgpool to fail liveness probe and subsequently restart as part of a normal and proper fail over?

Yes, before upgrading to pgpool-II version 4.5.0 to 4.5.1, when using the bitnami/postgresql-ha chart with default values, short downtimes would be expected if the Postgresql primary node goes down.

Because requirements may vary from one user to another, we usually encourage users fine-tune pgpool-II settings, repmgr settings and liveness/readiness timers to reduce those downtimes adapted to their use case.

The bug we are addressing in this issue is the behavior OP reported, where on primary failure pgpool-II would never establish a connection to the primary node and all client connections would fail until a manual restart is performed, which is a much severe issue.

**Comment by Stevenpc3** [comment]
Yes I agree it's a severe issue, we have that issue too which is why I asked 

"How did you test before publishing" and not see this? Was fail over part of the normal test? 

As for the pgpool liveness probe, you added a full restart of pgpool as normal operation. This is off to me. Would you expect pgpool to be restarted in a bare metal deploy? Why is pgpool expected to restart in kubernetes?

**Comment by Stevenpc3** [comment]
bump

**Comment by migruiz4** [comment]
I have created an internal task to fix this issue and added the on-hold label to prevent the stale bot while this is still relevant.

I'm sorry but we are a small team and have to deal with other priorities at the moment, if you would like to help us find a fix for this issue and contribute to the project, please send a pull request and we will be happy to review it.

**Comment by Stevenpc3** [comment]
Our team has been testing other solutions and holding still on a very old 9.x.x version of the charts we tried with some tweaks. 

The only consistent thing we found is that pgpool 4.5.1 appears to cause issues regardless. We had these issues pop up when using that version in the new charts and the older charts. So the issue must be pgpool itself or the use of it with the charts.

**Comment by Kavantix** [comment]
I believe I also just ran into this issue while testing out this chart.

Basically what happens is that pgpool seems to remember what the primary node was and when it then connects to that node again it is no longer the primary but it still thinks it is causing any writes to fail.

If however a write is issues while the old primary is not reachable pgpool will notice that and search for a new primary

**Comment by rafariossaa** [comment]
Hi,
  This is what I am trying to reproduce the issue:
  I have installed chart version `13.6.1` that uses pgpool version `4.5.0-debian-12-r9`
```bash
$ helm install mypg bitnami/postgresql-ha --version=13.6.1 --set postgresql.password=adminpwd --set postgresql.repmgrPassword=repmgrpwd --set pgpool.adminPassword=pgpoolpwd 
```
 Then upgrade the deployment to `14.0.13` (pgpool `4.5.1-debian-12-5`). As repgmr is updated, an extra step is needed in order to repmgr be able to be upgrade.

```bash
$ helm upgrade mypg bitnami/postgresql-ha --version=14.0.13 --set postgresql.password=adminpwd --set postgresql.repmgrPassword=repmgrpwd --set pgpool.adminPassword=pgpoolpwd --set postgresql.upgradeRepmgrExtension=true --set postgresql.replicaCount=1

$ helm upgrade mypg bitnami/postgresql-ha --version=14.0.13 --set postgresql.password=adminpwd --set postgresql.repmgrPassword=repmgrpwd --set pgpool.adminPassword=pgpoolpwd --set postgresql.replicaCount=3
```

Then, I checked the status of the cluster:
```bash
kubectl exec -it mypg-postgresql-ha-postgresql-0 -- /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf daemon status
...
 ID | Name                            | Role    | Status    | Upstream                        | repmgrd | PID | Paused? | Upstream last seent
----+---------------------------------+---------+-----------+---------------------------------+---------+-----+---------+---------------------
 1000 | mypg-postgresql-ha-postgresql-0 | primary | * running |                                 | running | 1   | no      | n/a
 1001 | mypg-postgresql-ha-postgresql-1 | standby |   running | mypg-postgresql-ha-postgresql-0 | running | 1   | no      | 1 second(s) ago    g
 1002 | mypg-postgresql-ha-postgresql-2 | standby |   running | mypg-postgresql-ha-postgresql-0 | running | 1   | no      | 0 second(s) ago
```

So, in this case, the primary node is `postgresql-0`. I deployed a new pod, to insert some data meanwhile testing.

```bash
$ kubectl run -it pgclient --image=bitnami/minideb -- bash
$ apt-get update; apt-get install postgresql-client -y
$ export PGPASSWORD=adminpwd
$ echo "create table test (c1 varchar(100));" | psql -U postgres -h  mypg-postgresql-ha-pgpool
$ echo "insert into test values ('1234'); \watch" | psql -U postgres -h  mypg-postgresql-ha-pgpool
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
...
```

Now, If I kill the primary node:
```bash
$ kubectl delete  pod/mypg-postgresql-ha-postgresql-0
```
The primary is now `postgresql-1`
```
kubectl exec -it mypg-postgresql-ha-postgresql-0 -- /opt/bitnami/scripts/postgresql-repmgr/entrypoint.sh repmgr -f /opt/bitnami/repmgr/conf/repmgr.conf daemon status
...
...

 ID | Name                            | Role    | Status    | Upstream                        | repmgrd | PID | Paused? | Upstream last seen 
----+---------------------------------+---------+-----------+---------------------------------+---------+-----+---------+--------------------
 1000 | mypg-postgresql-ha-postgresql-0 | standby |   running | mypg-postgresql-ha-postgresql-1 | running | 1   | no      | 0 second(s) ago
 1001 | mypg-postgresql-ha-postgresql-1 | primary | * running |                                 | running | 1   | no      | n/a
 1002 | mypg-postgresql-ha-postgresql-2 | standby |   running | mypg-postgresql-ha-postgresql-1 | running | 1   | no      | 1 second(s) ago
```

  The inserting command, aborted when the connection was closed, but when reruning it again, it continues inserting data without issues.

  Could you give this a try ? Please, write down any steps you do because this kind of issues are tricky to reproduce and I would like to have a step sequence to reproduce it.

**Comment by Stevenpc3** [comment]
A small difference might be that this test does a constant insert and that might trigger the "notice it's down" that https://github.com/bitnami/charts/issues/23219#issuecomment-2155784928 mentioned.  In our use case we do not have a continuous write.  There is delays of minutes possibly between writes.  

A better test might be to bring up the postgresql instances, then write to the db once or twice.  Then stop writing.  Then delete the postgresql pod and wait a few seconds or so, then write to the db again to see if it happens instead of the constant 2 second writes watch is doing.

**Comment by Stevenpc3** [comment]
We had someone do a bit of testing...

We tried to reproduce the steps described in a comment by user "rafariossaa" (https://github.com/bitnami/charts/issues/23219#issuecomment-2165749114).
 
Summary:
* The psql command from rafariossaa generally does not reproduce the pgpool issue.
* We wrote a different psql command that generally does cause the pgpool issue (against the current beast configuration).
* On recommendation, we tested against the current "vanilla" bitnami postgresql-ha (14.2.7) and were able to reproduce the problem again.
* Setting some pgpool config items (connection_cache = false, failover_on_backend_error = on, failover_on_backend_shutdown = on) against the 14.2.7 version seems to fix it.
 
The longer version:
 
The comment from rafariossaa sets up the database, adds a separate pod to run psql from (which is a good trick), and then runs a command to connect to pgpool and continuously insert into a test table. The watch is inside psql, so there's one db connection that is kept open.
echo "insert into test values ('1234'); \watch" | psql -U postgres -h  mypg-postgresql-ha-pgpool
Generally, failing the primary pod here causes the psql connection to break and you can restart the command and things work seem fine.
 
You can change the psql command so that it creates a db connection for each command (here, an insert and a select).
while true; do echo "insert into test values ('123'); select count(*) from test;" | psql -t -U postgres -h mypg-postgresql-ha-pgpool ; sleep 3; done
Generally, this reproduces the pgpool issue.  Failing the primary pod causes a few failed connections (connection refused while things fail over) and then an error about the read-only transaction ("ERROR:  cannot execute INSERT in a read-only transaction"), where pgpool has apparently failed to point to the correct postgres instance.
 
On suggestion, we installed the "vanilla" bitnami charts. In the as-is configuration, this will reproduce the read-only transaction issue.
 
After adding the settings from the Github thread, I've been unable to reproduce the issue so far:
pgpool:
  configuration: |-
   connection_cache = false
   failover_on_backend_error = on
   failover_on_backend_shutdown = on

Will do more testing.

**Comment by rafariossaa** [comment]
Hi,
  Thanks for your feedback.

**Comment by gtudan** [comment]
We are still experiencing this issue. Disabling the connection cache feels like a bad move, as it pretty much defeats the purpose of using a pooler in the first place

**Comment by rafariossaa** [comment]
@gtudan, consider that if you loose the master, you are going to have some connection cut/re-connection. Could you elaborate a bit more in you case or open a new issue ? In that case, feel free to link to this one if needed.

**Comment by gtudan** [comment]
@rafariossaa Sure, if the master is gone, the connection cache for that node has to be flushed. 

In our scenario we observed writes on the slave even when both nodes were up - I _assume_ that happened after a helm upgrade when a failover/failback was not properly performed, but I'm not sure. 

I totally understand that disabling the connection cache prevents those issues, as the connections are reestablished every time. But I don't quite see why the cache should have to be disabled in normal operations to prevent write operations ending up on the slave. 

Unfortunately, we had to rollback to the non-ha chart, because we could not figure out this issue, so I'm currently not able to reproduce this. We'll plan to retry in a few weeks, hopefully I'll be able to provide further insights then.

**Comment by rafariossaa** [comment]
I don't like the idea of disable the cache either.
Thanks, don't hesitate to share your findings.

**Comment by yukha-dw** [comment]
> You can add the flag
> 
> ```
> postgresql-ha:
>   pgpool:
>     configuration: |
>       connection_cache = false
>       failover_on_backend_error = on
>       failover_on_backend_shutdown = on
> ```

This config does not working for me. One of 3 Pgpool nodes keeps setting previously replica node as primary node when doing `helm upgrade`.

**Comment by likeMyCoffee** [comment]
For me the issue was that my application was connecting to the postgresql headless service instead of pgpool service. This meant I was going directly to a random postgresql server, possibly a standby one, giving the transaction error.

Hope this kinda helps someone…

**Comment by juan131** [comment]
Hi everybody! I think this issue should be solved after https://github.com/bitnami/charts/pull/30676 thanks to the contribution @seatrain did at https://github.com/bitnami/containers/pull/75225

**Comment by yukha-dw** [comment]
> Hi everybody! I think this issue should be solved after #30676 thanks to the contribution @seatrain did at [bitnami/containers#75225](https://github.com/bitnami/containers/pull/75225)

I think so, this solution fix most of read-only issues for me.

If anyone interested, we can open another issue to fix case when we do `helm upgrade`. Because how helm works, it will redeploy `Pgpool` and `PostgreSQL` altogether, without waiting for `PostgreSQL` to be ready. This sometimes causes `Pgpool` to have outdated node pool configuration and stuck with it.

**Comment by juan131** [comment]
Yes please @yukha-dw, I'd appreciate if you could create a different issue for that topic and close this one.
